File: dotnet/src/PqcStandards/SlhDsa/SlhParams.cs
namespace PqcStandards.SlhDsa;

/// <summary>SLH-DSA parameter sets per FIPS 205.</summary>
public sealed record SlhParams(
    string Name,
    int N,          // Security parameter (hash output bytes)
    int H,          // Total tree height
    int D,          // Number of layers (hypertree layers)
    int HPrime,     // Height of each tree = H/D
    int A,          // FORS trees count
    int K,          // FORS leaves per tree (log)
    int W,          // Winternitz parameter
    int Len,        // WOTS+ chain count
    string HashFamily,  // "SHAKE" or "SHA2"
    int SigSize)
{
    // SHAKE-based
    public static readonly SlhParams Shake128s = new("SLH-DSA-SHAKE-128s", 16, 63, 7, 9, 12, 14, 16, 35, "SHAKE", 7856);
    public static readonly SlhParams Shake128f = new("SLH-DSA-SHAKE-128f", 16, 66, 22, 3, 6, 33, 16, 35, "SHAKE", 17088);
    public static readonly SlhParams Shake192s = new("SLH-DSA-SHAKE-192s", 24, 63, 7, 9, 14, 17, 16, 51, "SHAKE", 16224);
    public static readonly SlhParams Shake192f = new("SLH-DSA-SHAKE-192f", 24, 66, 22, 3, 8, 33, 16, 51, "SHAKE", 35664);
    public static readonly SlhParams Shake256s = new("SLH-DSA-SHAKE-256s", 32, 64, 8, 8, 14, 22, 16, 67, "SHAKE", 29792);
    public static readonly SlhParams Shake256f = new("SLH-DSA-SHAKE-256f", 32, 68, 17, 4, 9, 35, 16, 67, "SHAKE", 49856);

    // SHA2-based
    public static readonly SlhParams Sha2_128s = new("SLH-DSA-SHA2-128s", 16, 63, 7, 9, 12, 14, 16, 35, "SHA2", 7856);
    public static readonly SlhParams Sha2_128f = new("SLH-DSA-SHA2-128f", 16, 66, 22, 3, 6, 33, 16, 35, "SHA2", 17088);
    public static readonly SlhParams Sha2_192s = new("SLH-DSA-SHA2-192s", 24, 63, 7, 9, 14, 17, 16, 51, "SHA2", 16224);
    public static readonly SlhParams Sha2_192f = new("SLH-DSA-SHA2-192f", 24, 66, 22, 3, 8, 33, 16, 51, "SHA2", 35664);
    public static readonly SlhParams Sha2_256s = new("SLH-DSA-SHA2-256s", 32, 64, 8, 8, 14, 22, 16, 67, "SHA2", 29792);
    public static readonly SlhParams Sha2_256f = new("SLH-DSA-SHA2-256f", 32, 68, 17, 4, 9, 35, 16, 67, "SHA2", 49856);

    /// <summary>All parameter sets.</summary>
    public static readonly SlhParams[] All = [
        Shake128s, Shake128f, Shake192s, Shake192f, Shake256s, Shake256f,
        Sha2_128s, Sha2_128f, Sha2_192s, Sha2_192f, Sha2_256s, Sha2_256f
    ];
}

File: java/src/main/java/com/pqc/slhdsa/SlhParams.java
package com.pqc.slhdsa;

/**
 * SLH-DSA parameter sets per FIPS 205.
 * All 12 parameter sets: {SHAKE,SHA2} x {128,192,256} x {f,s}.
 */
public enum SlhParams {

    // SHAKE-based parameter sets (FIPS 205 Table 1)
    // Constructor: (name, n, w, m_unused, hPrime, d, a, k, lgT_unused, hashFamily)
    SHAKE_128F("SLH-DSA-SHAKE-128f", 16, 16, 0,  3, 22,  6, 33, 0, HashFamily.SHAKE),
    SHAKE_128S("SLH-DSA-SHAKE-128s", 16, 16, 0,  9,  7, 12, 14, 0, HashFamily.SHAKE),
    SHAKE_192F("SLH-DSA-SHAKE-192f", 24, 16, 0,  3, 22,  8, 33, 0, HashFamily.SHAKE),
    SHAKE_192S("SLH-DSA-SHAKE-192s", 24, 16, 0,  9,  7, 14, 17, 0, HashFamily.SHAKE),
    SHAKE_256F("SLH-DSA-SHAKE-256f", 32, 16, 0,  4, 17,  9, 35, 0, HashFamily.SHAKE),
    SHAKE_256S("SLH-DSA-SHAKE-256s", 32, 16, 0,  8,  8, 14, 22, 0, HashFamily.SHAKE),

    // SHA2-based parameter sets (FIPS 205 Table 1)
    SHA2_128F("SLH-DSA-SHA2-128f", 16, 16, 0,  3, 22,  6, 33, 0, HashFamily.SHA2),
    SHA2_128S("SLH-DSA-SHA2-128s", 16, 16, 0,  9,  7, 12, 14, 0, HashFamily.SHA2),
    SHA2_192F("SLH-DSA-SHA2-192f", 24, 16, 0,  3, 22,  8, 33, 0, HashFamily.SHA2),
    SHA2_192S("SLH-DSA-SHA2-192s", 24, 16, 0,  9,  7, 14, 17, 0, HashFamily.SHA2),
    SHA2_256F("SLH-DSA-SHA2-256f", 32, 16, 0,  4, 17,  9, 35, 0, HashFamily.SHA2),
    SHA2_256S("SLH-DSA-SHA2-256s", 32, 16, 0,  8,  8, 14, 22, 0, HashFamily.SHA2);

    public enum HashFamily { SHAKE, SHA2 }

    public final String name;
    /** Security parameter (bytes): n */
    public final int n;
    /** Winternitz parameter w (always 16 for FIPS 205) */
    public final int w;
    /** FORS message length in bits: a * k */
    public final int m;
    /** Height of each XMSS tree: hPrime */
    public final int hPrime;
    /** Number of XMSS layers: d */
    public final int d;
    /** Total tree height: h = hPrime * d */
    public final int h;
    /** FORS trees: k */
    public final int k;
    /** FORS tree height: a */
    public final int a;
    /** Hash family */
    public final HashFamily hashFamily;

    /** WOTS+ chain length: w - 1 = 15 */
    public final int wotsLen1;
    /** WOTS+ checksum length */
    public final int wotsLen2;
    /** WOTS+ total length: len1 + len2 */
    public final int wotsLen;
    /** log2(w) = 4 */
    public final int lgW;

    SlhParams(String name, int n, int w, int m, int hPrime, int d, int a, int k, int lgT,
              HashFamily hashFamily) {
        // m here is unused in the enum spec; we compute from a, k
        this.name = name;
        this.n = n;
        this.w = w;
        this.hPrime = hPrime;
        this.d = d;
        this.h = hPrime * d;
        this.a = a;
        this.k = k;
        this.hashFamily = hashFamily;
        this.lgW = 4; // log2(16) = 4

        // WOTS+ parameters per FIPS 205 Section 5
        this.wotsLen1 = (8 * n + lgW - 1) / lgW; // ceil(8n / lgW)
        // len2 = floor(log_w(len1 * (w-1))) + 1
        int maxChecksum = wotsLen1 * (w - 1);
        int tmp = 0;
        int val = maxChecksum;
        while (val > 0) {
            val /= w;
            tmp++;
        }
        this.wotsLen2 = tmp;
        this.wotsLen = wotsLen1 + wotsLen2;
        this.m = a * k;
    }

    /** Signature size in bytes. */
    public int sigSize() {
        // SLH-DSA sig = (randomizer n) + (FORS sig: k*(a+1)*n) + (HT sig: d*(hPrime + wotsLen)*n)
        return n + (k * (a + 1) * n) + (d * (hPrime + wotsLen) * n);
    }

    /** Public key size = 2n bytes (PK.seed || PK.root). */
    public int pkSize() {
        return 2 * n;
    }

    /** Secret key size = 4n bytes (SK.seed || SK.prf || PK.seed || PK.root). */
    public int skSize() {
        return 4 * n;
    }

    public static SlhParams fromName(String name) {
        for (SlhParams p : values()) {
            if (p.name.equalsIgnoreCase(name)) return p;
        }
        throw new IllegalArgumentException("Unknown SLH-DSA parameter set: " + name);
    }
}
